internal/cmd/agent/controllers/bundledeployment/controller.go

```diff
@@ -130,11 +136,110 @@ func (h *handler) DeployBundle(bd *fleet.BundleDeployment, status fleet.BundleDe
 	status.Release = release
 	status.AppliedDeploymentID = bd.Spec.DeploymentID
 
+	err = h.setNamespaceLabelsAndAnnotations(bd, release)
+
+	if err != nil {
+		return fleet.BundleDeploymentStatus{}, err
+	}
+
 	// Setting the error to nil clears any existing error
 	condition.Cond(fleet.BundleDeploymentConditionInstalled).SetError(&status, "", nil)
 	return status, nil
 }
 
+func (h *handler) setNamespaceLabelsAndAnnotations(bd *fleet.BundleDeployment, releaseID string) error {
+	if bd.Spec.Options.NamespaceLabels == nil && bd.Spec.Options.NamespaceAnnotations == nil {
+		return nil
+	}
+
+	ns, err := h.fetchNamespace(releaseID)
+	if err != nil {
+		return err
+	}
+
+	if reflect.DeepEqual(bd.Spec.Options.NamespaceLabels, ns.Labels) && reflect.DeepEqual(bd.Spec.Options.NamespaceAnnotations, ns.Annotations) {
+		return nil
+	}
+
+	if ns.Annotations == nil && bd.Spec.Options.NamespaceAnnotations != nil {
+		ns.Annotations = map[string]string{}
+	}
+	if bd.Spec.Options.NamespaceLabels != nil {
+		addLabelsFromOptions(ns.Labels, *bd.Spec.Options.NamespaceLabels)
+	}
+	if bd.Spec.Options.NamespaceAnnotations != nil {
+		addAnnotationsFromOptions(ns.Annotations, *bd.Spec.Options.NamespaceAnnotations)
+	}
```

Comment:
I feel that this would be easier to read. How do you see it?
```suggestion
	if bd.Spec.Options.NamespaceLabels != nil {
		addLabelsFromOptions(ns.Labels, *bd.Spec.Options.NamespaceLabels)
	}
	if bd.Spec.Options.NamespaceAnnotations != nil {
	        if ns.Annotations == nil {
	                ns.Annotations = map[string]string{}
	       }
		addAnnotationsFromOptions(ns.Annotations, *bd.Spec.Options.NamespaceAnnotations)
	}
```